File: 152 Maximum Product Subarray/152.java
class Solution {
    public int maxProduct(int[] nums) {
        int preMax = 1, preMin = 1;
        int curMax = 0, curMin = 0;
        int globalMax = Integer.MIN_VALUE;
        for (int i : nums) {
            if (i < 0) {
                curMax = Math.max(i, preMin * i);
                curMin = Math.min(i, preMax * i);
            } else {
                curMax = Math.max(i, preMax * i);
                curMin = Math.min(i, preMin * i);
            }
            preMax = curMax;
            preMin = curMin;
            globalMax = Math.max(globalMax, preMax);
        }
        return globalMax;
    }

    // Time Complexity: O(n)
    // Space Complexity: O(1)
}

File: 152 Maximum Product Subarray/152.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        globalMax = -float('inf')
        curMax, curMin = None, None
        preMax, preMin = 1, 1
        for i in nums:
            if i < 0:
                curMax = max(i, i * preMin)
                curMin = min(i, i * preMax)
            else:
                curMax = max(i, i * preMax)
                curMin = min(i, i * preMin)
            preMax = curMax
            preMin = curMin
            globalMax = max(globalMax, preMax)
        return globalMax